cpp/src/arrow/compute/kernels/scalar_nested_test.cc

```diff
@@ -141,16 +148,25 @@ TEST(TestScalarNested, StructField) {
                 &extract0);
     CheckScalar("struct_field", {arr}, ArrayFromJSON(int64(), "[10, 11, 12, null]"),
                 &extract20);
+
+    CheckScalar("struct_field", {arr}, ArrayFromJSON(int32(), "[1, null, 3, null]"),
+                &extract0_field_ref_path);
+    CheckScalar("struct_field", {arr}, ArrayFromJSON(int32(), "[1, null, 3, null]"),
+                &extract0_field_ref_name);
+    CheckScalar("struct_field", {arr}, ArrayFromJSON(int64(), "[10, 11, 12, null]"),
+                &extract20_field_ref_nest);
+
     EXPECT_RAISES_WITH_MESSAGE_THAT(Invalid,
                                     ::testing::HasSubstr("out-of-bounds field reference"),
                                     CallFunction("struct_field", {arr}, &invalid1));
     EXPECT_RAISES_WITH_MESSAGE_THAT(Invalid,
-                                    ::testing::HasSubstr("out-of-bounds field reference"),
+                                    ::testing::HasSubstr("No match for FieldRef"),
```

Comment:
Ya sorry, this was just a sub-string to the whole error, which is:
> No match for FieldRef.FieldPath(2 4) in struct<a: int32, b: string, c: struct<d: int64, e: double>>

Is this okay, and/or should I include the whole/more of it into the `HasSubstr`check?

Reply:
My comment wasn't really about only checking a substring. But that the former error message (which included "out-of-bounds field reference") seems more informative as "No match for FieldRef.FieldPath(2 4) in ...")

So my question is not to update the HasSubtr test line here, but rather: is it possible / would it be worth to try to update `FieldRef::FindOne` (where the error comes from now, I think) to also give such a more informative error message (so we don't loose this more informative error message)

(this might not necessarily be a straightforward fix, didn't check how `FindOne` currently does error checking, so even if we would like to update this, it could also be for a follow-up JIRA)

Reply:
Hmm, the latter message actually seems more informative to me, if it includes the entire FieldRed and struct type string representation. But I don't know exactly how the former error message looked :-)

Reply:
The former error message was like this: 
`out-of-bounds field reference to field 4 in type struct<a: int32, b: string, c: struct<d: int64, e: double>> with 3 fields`

@jorisvandenbossche Should have made it clearer that I did understand your comment, I only wanted to confirm you were aware the 'new' error did contain additional context and if that was as good as the previous. I'll look more deeply at how to retain the previous error if that's preferable, I don't have much of an opinion on this.

Reply:
Looked around and tried a bit, but the point at which the context of the error is being created in `FieldRef::FindAll(const FieldVector& fields) const` no longer has the context of the `DataType` and thus cannot directly discover `type.num_fields()` as `CheckIndex` does. May be possible return the `out_of_range_depth` variable there back out to the call `FindOne(DataType)` in a error status.

Anyhow, I'd happy to open a JIRA and take it as a follow up if that's desired. :+1: 

A summary of where the new error is coming from:
```cpp
  FieldPath field_path;
  if (field_ref.IsNested() || field_ref.IsName()) {
    // New error happens here in FindOne
    ARROW_ASSIGN_OR_RAISE(field_path, field_ref.FindOne(*type));
  } else {
    field_path = *field_ref.field_path();
  }

  for (const auto& index : field_path.indices()) {
    // Ideally, we get the error from here, CheckIndex
    RETURN_NOT_OK(StructFieldFunctor::CheckIndex(index, *type));
    type = type->field(index)->type().get();
  }
```

Reply:
> Anyhow, I'd happy to open a JIRA and take it as a follow up if that's desired. +1

Yes, let's file a JIRA for this

Reply:
[ARROW-18295](https://issues.apache.org/jira/browse/ARROW-18295) :ok_hand: